src/protocols/bdx/BdxTransferSession.cpp

```diff
@@ -0,0 +1,873 @@
+}
+
+/**
+ * @brief
+ *   Allocate a new PacketBuffer and write data from a BDX message struct.
+ */
+template <class BdxMsgType>
+CHIP_ERROR WriteToPacketBuffer(const BdxMsgType & msgStruct, ::chip::System::PacketBufferHandle & msgBuf)
+{
+    CHIP_ERROR err     = CHIP_NO_ERROR;
+    size_t msgDataSize = msgStruct.MessageSize();
+    msgBuf             = ::chip::System::PacketBufferHandle::New(static_cast<uint16_t>(msgDataSize));
+    if (msgBuf.IsNull())
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    ::chip::BufBound bbuf(msgBuf->Start(), msgBuf->AvailableDataLength());
+    msgStruct.WriteToBuffer(bbuf);
+    msgBuf->SetDataLength(static_cast<uint16_t>(bbuf.Needed()));
+
+    return err;
+}
+
+namespace chip {
+namespace bdx {
+
+TransferSession::TransferSession()
+{
+    mState         = kState_Idle;
+    mPendingOutput = kOutput_None;
+    mSuppportedXferOpts.SetRaw(0);
+}
+
+void TransferSession::PollOutput(OutputEvent & event, uint64_t curTimeMs)
+{
+    event = OutputEvent();
+
+    if (mAwaitingResponse && (mTimeoutStartTimeMs + mTimeoutMs <= curTimeMs))
+    {
+        event             = OutputEvent(kOutput_TransferTimeout);
+        mState            = kState_Error;
+        mAwaitingResponse = false;
+        return;
+    }
+
+    switch (mPendingOutput)
+    {
+    case kOutput_None:
+        event = OutputEvent(kOutput_None);
+        break;
+    case kOutput_InternalError:
+        event = OutputEvent::StatusReportEvent(kOutput_InternalError, mStatusReportData);
+        break;
+    case kOutput_StatusReceived:
+        event = OutputEvent::StatusReportEvent(kOutput_StatusReceived, mStatusReportData);
+        break;
+    case kOutput_MsgToSend:
+        event               = OutputEvent(kOutput_MsgToSend);
+        event.MsgData       = std::move(mPendingMsgHandle);
+        mTimeoutStartTimeMs = curTimeMs;
+        break;
+    case kOutput_InitReceived:
+        event = OutputEvent::TransferInitEvent(mTransferRequestData, std::move(mPendingMsgHandle));
+        break;
+    case kOutput_AcceptReceived:
+        event = OutputEvent::TransferAcceptEvent(mTransferAcceptData, std::move(mPendingMsgHandle));
+        break;
+    case kOutput_QueryReceived:
+        event = OutputEvent(kOutput_QueryReceived);
+        break;
+    case kOutput_BlockReceived:
+        event = OutputEvent::BlockDataEvent(mBlockEventData, std::move(mPendingMsgHandle));
+        break;
+    case kOutput_AckReceived:
+        event = OutputEvent(kOutput_AckReceived);
+        break;
+    case kOutput_AckEOFReceived:
+        event = OutputEvent(kOutput_AckEOFReceived);
+        break;
+    default:
+        event = OutputEvent(kOutput_None);
+        break;
+    }
+
+    mPendingOutput = kOutput_None;
+}
+
+CHIP_ERROR TransferSession::StartTransfer(TransferRole role, const TransferInitData & initData, uint32_t timeoutMs,
+                                          uint64_t curTimeMs)
```

Comment:
I added this to initialize `mTimeoutStartTimeMs`, because otherwise it will be initialized to 0 and the first call to PollOutput() will check for the timeout and do `curTimeMs - 0 > mTimeoutMs` which is incorrect.

Admittedly this is probably leaking implementation details. I can add something to initialize `mTimeoutStartTimeMs` on the first call to PollOutput()